### ISSUE TITLE
Installation design flaw

### ISSUE THREAD

**Issue by eddelbuettel** [astamm/nloptr#87]
First off, thanks for looking after nloptr and (hopefully) ensuring it remains on CRAN.

Now, my system already has a well taken-care-of installation of NLopt:

```sh
edd@rob:~/git/nloptr(master)$ dpkg -l| grep nlopt | cut -c-70
ii  libnlopt-dev:amd64                         2.7.0-4                
ii  libnlopt0:amd64                            2.7.0-4                
edd@rob:~/git/nloptr(master)$ 
```

It is wasteful (though, sadly, not uncommon) to unconditionally install dependencies.  I would urge to consider deploying existing installations.  _Hundreds_ of CRAN packages can do that, it would be nice if nloptr did too.  And that is before we get into semi-religious debates about static and shared linking...

```sh
edd@rob:~/git/nloptr(master)$ pkg-config --list-all | grep nlopt
nlopt                          NLopt - nonlinear optimization libary
edd@rob:~/git/nloptr(master)$ pkg-config --cflags nlopt
                   
edd@rob:~/git/nloptr(master)$ pkg-config --libs nlopt
-lnlopt
edd@rob:~/git/nloptr(master)$ 
````

So we see that the compiler will find the headers by default, and all we need is adding one link instruction.  It would be nice if you could restore the ability to use an existing installation.

**Comment by eddelbuettel** [comment]
Adding explicit CC to @astamm in case I filed this at the wrong repo.  You can send email to GH to request yours to be the main one; I just followed my prior practice where this was the parentrepo.  Happy to adjust....

**Comment by eddelbuettel** [comment]
Oh, and by the way, a simple example of allowing both installation methods is my [RcppRedis](https://github.com/eddelbuettel/rcppredis) package I happen to be working on a little this week -- it defaults to using an external libhiredis and if none is found it builds from the included sources.

**Comment by astamm** [comment]
Thanks for this discussion and the link to [RcppRedis](https://github.com/eddelbuettel/rcppredis). I will look into that as well.

**Comment by eddelbuettel** [comment]
I actually have a few more packages like that (I could grep some more ...) and Jeroen has a large number too where he (as an `autoconf` hater) calls the configure script `anticonf`.  

Because of his somewhat obnoxious naming, that is easily searchable:   https://github.com/search?q=org%3Acran+anticonf  gets me 41 hits.

Your attempt of relying on `cmake` is of course defensible.  As is my insistence to not compile when we do not have to :)   

What I would try next is a (simple, as e.g. in RcppRedis) configure script (which you could write 'anticonf' style) that 
- looks for `pkg-config` 
- if found uses it to learn whether nlopt is installed (`pkg-config --exists nlopt`) 
- (if you insist you could even check minimum versions but `nloptr` should not need that)
- if present extract compiler and linker flags and use them
to give the fastest possible installation on Linux and possibly macOS with proper `brew` setup

If all that came up negative, check for `cmake` and install from source if present.   That could also be the default on Windows and macOS as users typically get a binary from CRAN anyway and the unconditional compilation a) done at CRAN and b) amortized over many users.

Bing me if you're lost is some details. I have been working on this very aspect of building with external libraries at CRAN for 20 years.  Feel free to take any and all advantage of that.

**Comment by eddelbuettel** [comment]
Any progress?  I would hate for this to be thrown off CRAN for a logical flaw in

> The NLopt library has switched to CMake for its build system. Consequently, we updated the nloptr package accordingly.

Upstream build systems have next to nothing to do with CRAN build systems.  There are over 18,000 package, over 4400 packages with compiled --- but last time I counted _less than ten that use CMake_.   I fear you are boxing yourself into a corner here.  I can help.

If not, I'll just fork.  It's not the end of the world either.

**Comment by astamm** [comment]
The current version (on master) successfully passes checks on winbuilder release and devel, rhub, RStudio GitHub action script for testing on macOS, windows and Linux and locally on my Mac.

So I submitted it before Christmas and it went back with pre test error which, for what I could gather, pertains to the current release and not the submitted one.

Hence, I feel there is no reason for the version on the master not to be accepted by CRAN.

CRAN members however have not yet replied back at me despite two reminders (last of which was yesterday).

I will take care after that (probably indeed with your help) to put back the ability to use an existing version of nlopt, but I am quite not sure it is mandatory to keep nloptr from being archived.

**Comment by astamm** [comment]
Of course, if you think it would take you very little of your time to add back this ability while keeping the cmake way in case nlopt has to built from included source, I would be grateful. But I understand if you cannot and this is something I will look into as soon as CRAN answers to my email.

**Comment by eddelbuettel** [comment]
It is easy _if_ we do the following
- keep your cmake file and documentation
- make it _optional_ i.e. restore the top-level `configure.ac` creating a `configure` (side advantage: no longer a _hard_ dependency on `cmake` on Linux)
- when `configure` determines a build is needed it builds via `cmake`

I can set that up fairly easily -- the [tiledb](https://github.com/TileDB-Inc/TileDB-R) package I look after (for my day job) does just that in [its configure.ac](https://github.com/TileDB-Inc/TileDB-R/blob/master/configure.ac) (it actually does one more of accepting a non-standard location, we can skip that).

If that scheme is acceptable to you I can work on it.  

(As for CRAN: when I did `R CMD build` and `R CMD CRAN` a few days ago it built but still emitted tons of warnings. It seemed to me at the time that CRAN would not take a package in that state.  I can be wrong, despite nearly 20 years of working with them they can be hard to read.)

**Comment by astamm** [comment]
That scheme looks pretty well for me. I would be grateful if you could contribute indeed. I am discussing in parallel with CRAN folks about accepting it in its current state. It seems the pre test errors only because it cannot find `cmake` on the win-builder machine. But indeed once this step is resolved, there might be other warnings they uncover on other operating systems despite my successful checks on r-hub and win-builder and all.

So I think you are quite right and we should prepare for adding back the ability to use a system build of `nlopt` in case there is one, which should smooth things especially on Linux machines.

I will hardly be able to write code a lot this afternoon and tomorrow as I am traveling. But I am happy to read and answer as much as necessary to get this done.

Thanks very much.

**Comment by eddelbuettel** [comment]
I just re-ran `R CMD build` and `R CMD check`.  Maybe my memory was sloppy -- the warnings only come from compilation of the external code (you chose to have a hard depend on) so they will nag you about this but it shouldn't be a show stopper.   As you insist on a hard depend, you could 'vendor it' i.e. include the source and adjust them. I do that, grudingly, for package [BH](https://cloud.r-project.org/web/packages/BH/index.html) (with patches [here](https://github.com/eddelbuettel/bh/tree/master/local/patches)).

And yes, a hard system depends on `cmake` for Windows (and macOS) may actually be quite defensible as it frees them from having to build the library themselves (and keep it current).

And yes indeed an optional skipping of forced builds would be nicer. I will try to work on this this evening.

**Comment by astamm** [comment]
Thanks for letting me know. The source files are included in the package though I haven't tweaked them. This was done mainly to avoid downloading them which could fail from time to time if the corresponding URL for some reason is temporarily down. I will let you know the answer from CRAN.

**Comment by eddelbuettel** [comment]
Nice. I agree that embedding the source is more robust.  I started to do that in a few packages too:  RcppRedis require libhiredis, now it ships it and builds as needed;  in RDieHarder I used to use an external library and gave up on it (I am the de-factor maintainer of it and now embed).  TileDB is too big to embed.

This is all good as you seem to have upstream's ear.  So these (somewhat cosmetic) warnings may in time be addressed.  First priority is to get nloptr back into CRAN's good graces.  That is more important that my nagging on using a (more common) external libnlopt where available.  Maybe we get both done.  I'll do some work later.

**Comment by eddelbuettel** [comment]
There is something else going with M1 builds that fail -- that is back over to you as I can't seem to get the actual installation log from RHub (or have forgotten how-to).  I need to concentrate on something else.

I did fix your use/non-use of ARCH though and can add this to the PR if your're interested.

**Comment by astamm** [comment]
Thank you for pointing that out. I will run it on r-hub with this platform.

And thanks a lot for putting back the ability to use an already existing build of `nlopt`. I am more than willing to receive your PR.

Many thanks.

**Comment by eddelbuettel** [comment]
We'll get there. 

And `cmake` itself can do that test for an existing library too, I can ask a colleague for help on that (as I am not so fluent with `cmake`) so that in a few weeks/month we can move away from autoconf + cmake to just cmake (if it is indeed present everywhere, I still remain somewhat divided on that).

First priority is to not get thrown off CRAN.  Rope me in if you need help with test results or emails or ...

**Comment by astamm** [comment]
Here is where the package fails to install on M1 Mac R-Hub platform:
```
set CMAKE_BIN=/Applications/CMake.app/Contents/bin/cmake
'getOption("repos")' replaces Bioconductor standard repositories, see
'?repositories' for details

replacement repositories:
    CRAN: https://cloud.r-project.org

set CC=clang -arch arm64
set CFLAGS=-I/opt/R/arm64/include -fPIC -falign-functions=64 -Wall -g -O2
set CXX=clang++ -arch arm64
set CXXFLAGS=-std=gnu++11 -I/opt/R/arm64/include -fPIC -falign-functions=64 -Wall -g -O2
set LDFLAGS=-L/opt/R/arm64/lib
./configure: line 29: 27616 Killed: 9               ${CMAKE_BIN} -D BUILD_SHARED_LIBS=OFF -D CMAKE_BUILD_TYPE=Release -D INSTALL_LIB_DIR=nlopt/lib -D CMAKE_AR=${AR} -D CMAKE_RANLIB=${RANLIB} -D NLOPT_CXX=ON -D NLOPT_GUILE=OFF -D NLOPT_MATLAB=OFF -D NLOPT_OCTAVE=OFF -D NLOPT_PYTHON=OFF -D NLOPT_SWIG=OFF -D NLOPT_TESTS=OFF -S nlopt-src -B nlopt-build
```

It is not clear to me what is going on here but I have an M1 iMac station in the office so I will test it locally there tomorrow afternoon.

In the meantime, I would not say that I have upstream's ear as they seem to to reply back at my email claiming the pre-test error is a false positive.

**Comment by eddelbuettel** [comment]
Yes I tickled the same issue a few times earlier but I am busy with work now. 

I think there was a way to get the `00Install.out` artifact from RHub but _right now_ I do not remember how. Can you try the alternate 'mac-builder' (though it may not `cmake`) over here:  https://mac.r-project.org/macbuilder/submit.html

**Comment by astamm** [comment]
Here is the link to the artifact: (https://artifacts.r-hub.io/nloptr_2.0.0.tar.gz-d2d8c7f2193c44e9865891d08278017d/nloptr.Rcheck/).

I'll try to run on the alternate 'Mac-builder' but not sure I manage to submit on the train, connexion is unstable.

**Comment by eddelbuettel** [comment]
Nice.  I think the artifact links are in the emails. I so far just worked from the direct html link from spawning the task.

Also, I'll probably make a like incremental commit to my PR.  That will hit a few of your shell scripts and just pass $ARCH around.  As we may or may not need it for M1.

**Comment by astamm** [comment]
Nice.

In the meantime, it did build successfully on the alternate Mac-builder: (https://mac.r-project.org/macbuilder/results/1642013491-a1e49ac854b44fa4/).

Not sure why it fails on r-hub M1 platform though. Maybe your PR will help.

**Comment by eddelbuettel** [comment]
Let's ignore RHub and M1 then because we have the existence proof that my PR does not do damage -- after all 'all it does' is call your CMake setup when needed.

Maybe my PR is good to go then?   (Note that I had to move one file to make room for my `configure` at the top.)

**Comment by astamm** [comment]
The version I tested on the alternate mac-builder is the one on my repo, not the one in the PR though.
But I think indeed it cannot do damage.
I will merge and run it again on the mac-builder just to be sure.

**Comment by eddelbuettel** [comment]
This has been sorted out, and fixed, in PR #88.

**Comment by eddelbuettel** [comment]
Ok -- fixing my own flaw as discussed per email.  _If_ a system NLopt is found but is too old, we now build too:

```sh
checking whether g++ -std=gnu++14 accepts -g... (cached) yes                              
checking for pkg-config... /usr/bin/pkg-config                                                           
checking if pkg-config knows NLopt... yes                                                                
checking for pkg-config checking NLopt version... insufficient: NLopt 2.7.0 or later is preferred.       
using NLopt via local cmake build on x86_64       
```

PR for that incoming in a moment.
